exporter/exporter.go

```diff
@@ -62,77 +66,123 @@ func New(opts *Opts) (*Exporter, error) {
 		client:           client,
-		collectors:       make([]prometheus.Collector, 0),
 		path:             opts.Path,
 		logger:           opts.Logger,
+		opts:             opts,
 		webListenAddress: opts.WebListenAddress,
 	}
 
-	if len(opts.CollStatsCollections) > 0 {
-		exp.collectors = append(exp.collectors, &collstatsCollector{
+	return exp, nil
+}
+
+func (e *Exporter) makeRegistry(ctx context.Context, client *mongo.Client) *prometheus.Registry {
+	registry := prometheus.NewRegistry()
+	if len(e.opts.CollStatsCollections) > 0 {
+		cc := collstatsCollector{
+			ctx:            ctx,
 			client:         client,
-			collections:    opts.CollStatsCollections,
-			compatibleMode: opts.CompatibleMode,
-			logger:         opts.Logger,
-		})
+			collections:    e.opts.CollStatsCollections,
+			compatibleMode: e.opts.CompatibleMode,
+			logger:         e.opts.Logger,
+		}
+		registry.MustRegister(&cc)
 	}
 
-	if len(opts.IndexStatsCollections) > 0 {
-		exp.collectors = append(exp.collectors, &indexstatsCollector{
+	if len(e.opts.IndexStatsCollections) > 0 {
+		ic := indexstatsCollector{
+			ctx:         ctx,
 			client:      client,
-			collections: opts.IndexStatsCollections,
-			logger:      opts.Logger,
-		})
+			collections: e.opts.IndexStatsCollections,
+			logger:      e.opts.Logger,
+		}
+		registry.MustRegister(&ic)
 	}
 
-	if !opts.DisableDiagnosticData {
-		exp.collectors = append(exp.collectors, &diagnosticDataCollector{
+	if !e.opts.DisableDiagnosticData {
+		ddc := diagnosticDataCollector{
+			ctx:            ctx,
 			client:         client,
-			compatibleMode: opts.CompatibleMode,
-			logger:         opts.Logger,
-		})
+			compatibleMode: e.opts.CompatibleMode,
+			logger:         e.opts.Logger,
+		}
+		registry.MustRegister(&ddc)
 	}
 
-	if !opts.DisableReplicasetStatus {
-		exp.collectors = append(exp.collectors, &replSetGetStatusCollector{
+	if !e.opts.DisableReplicasetStatus {
+		rsgsc := replSetGetStatusCollector{
+			ctx:            ctx,
 			client:         client,
-			compatibleMode: opts.CompatibleMode,
-			logger:         opts.Logger,
-		})
+			compatibleMode: e.opts.CompatibleMode,
+			logger:         e.opts.Logger,
+		}
+		registry.MustRegister(&rsgsc)
 	}
 
-	return exp, nil
+	return registry
 }
 
-// Run starts the exporter.
-func (e *Exporter) Run() {
-	registry := prometheus.NewRegistry()
-
-	for _, collector := range e.collectors {
-		registry.MustRegister(collector)
-	}
-
-	gatherers := prometheus.Gatherers{}
-	gatherers = append(gatherers, prometheus.DefaultGatherer)
-	gatherers = append(gatherers, registry)
+func (e *Exporter) handler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		client := e.client
+		// Use per-request connection.
+		if !e.opts.GlobalConnPool {
+			var err error
+			client, err = connect(ctx, e.opts.URI)
+			if err != nil {
+				e.logger.Errorf("Cannot connect to MongoDB: %v", err)
+				w.Header().Del(contentEncodingHeader)
```

Comment:
Why? At this point, nothing should be written

Reply:
Yes, right. Removed.